### PR TITLE
Add support for JFNK

### PR DESCRIPTION
**Pull request by avik-pal**

### CODE REVIEW

**Comment by avik-pal** [comment]
Ok definitely broke something :sweat:

File: src/jacobian.jl [hunk 1]
```diff
@@ -70,25 +74,34 @@ function build_jac_config(alg, f::F1, uf::F2, du1, u, tmp, du2) where {F1, F2}
...
             end
         end
     else
         jac_config = nothing
     end
-    jac_config
+
+    J = if !linsolve_needs_jac
+        # We don't need to construct the Jacobian
+        JacVec(uf, u; autodiff = AutoFiniteDiff())
```
Comment (ChrisRackauckas) [inline src/jacobian.jl:95]: this should follow the autodiff passed down in the algorithm

File: src/levenberg.jl [hunk 2]
```diff
@@ -228,21 +228,20 @@ end
...
+    du1 = zero(u)
+    du2 = zero(u)
+    tmp = zero(u)
+    J, jac_config = build_jac_and_jac_config(alg, f, uf, du1, u, tmp, du2)
 
     linprob = LinearProblem(J, _vec(zero(u)); u0 = _vec(zero(u)))
     weight = similar(u)
-    recursivefill!(weight, false)
+    # Q: Setting this to false leads to residual = 0 in GMRES?
+    recursivefill!(weight, true)
```
Comment (ChrisRackauckas) [inline src/levenberg.jl:240]:
```suggestion
    recursivefill!(weight, true)
```

yes, what was a bug, weight matrix should be identity by default not the zero matrix.

File: src/raphson.jl [hunk 3]
```diff
@@ -106,25 +106,20 @@ end
...
+    du1 = zero(u)
+    du2 = zero(u)
+    tmp = zero(u)
+    J, jac_config = build_jac_and_jac_config(alg, f, uf, du1, u, tmp, du2)
 
     linprob = LinearProblem(J, _vec(zero(u)); u0 = _vec(zero(u)))
     weight = similar(u)
-    recursivefill!(weight, false)
+    # Q: Setting this to false leads to residual = 0 in GMRES?
+    recursivefill!(weight, true)
```
Comment (ChrisRackauckas) [inline src/raphson.jl:118]:
```suggestion
    recursivefill!(weight, true)
```

File: src/trustRegion.jl [hunk 4]
```diff
@@ -280,21 +280,20 @@ end
...
+    du1 = zero(u)
+    du2 = zero(u)
+    tmp = zero(u)
+    J, jac_config = build_jac_and_jac_config(alg, f, uf, du1, u, tmp, du2)
 
     linprob = LinearProblem(J, _vec(zero(u)); u0 = _vec(zero(u)))
     weight = similar(u)
-    recursivefill!(weight, false)
+    # Q: Setting this to false leads to residual = 0 in GMRES?
+    recursivefill!(weight, true)
```
Comment (ChrisRackauckas) [inline src/trustRegion.jl:292]:
```suggestion
    recursivefill!(weight, true)
```

File: src/trustRegion.jl [hunk 5]
```diff
@@ -280,21 +280,20 @@ end
 
 function jacobian_caches(alg::TrustRegion, f, u, p, ::Val{true})
```
Comment (ChrisRackauckas) [inline src/trustRegion.jl:281]: Refactor this function so it's just one dispatch?

File: test/basictests.jl [hunk 6]
```diff
@@ -57,9 +58,11 @@ function ffiip(du, u, p)
 end
 u0 = [1.0, 1.0]
 
-sol = benchmark_inplace(ffiip, u0)
-@test sol.retcode === ReturnCode.Success
-@test all(abs.(sol.u .* sol.u .- 2) .< 1e-9)
+for linsolve in (nothing, KrylovJL_GMRES())
```
Comment (ChrisRackauckas) [inline test/basictests.jl:61]: add a test with a preconditioner?

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by avik-pal** [comment]
@ChrisRackauckas  can you rerun the integration test, it somehow got cancelled https://github.com/SciML/NonlinearSolve.jl/actions/runs/5941975798/job/16113962425?pr=201#step:6:1070

**Comment by ChrisRackauckas** [comment]
I think it's fine.

**Pull request merged**